v-next/core/src/internal/providers/http.ts

```diff
@@ -0,0 +1,361 @@
+import type {
+  EIP1193Provider,
+  RequestArguments,
+} from "../../types/providers.js";
+import type {
+  JsonRpcResponse,
+  JsonRpcRequest,
+  SuccessfulJsonRpcResponse,
+} from "../utils/json-rpc.js";
+import type {
+  Dispatcher,
+  RequestOptions,
+} from "@ignored/hardhat-vnext-utils/request";
+
+import EventEmitter from "node:events";
+import util from "node:util";
+
+import { HardhatError } from "@ignored/hardhat-vnext-errors";
+import { ensureError } from "@ignored/hardhat-vnext-utils/error";
+import { sleep, isObject } from "@ignored/hardhat-vnext-utils/lang";
+import {
+  getDispatcher,
+  isValidUrl,
+  postJsonRequest,
+  shouldUseProxy,
+  ConnectionRefusedError,
+  RequestTimeoutError,
+  ResponseStatusCodeError,
+} from "@ignored/hardhat-vnext-utils/request";
+
+import {
+  getJsonRpcRequest,
+  isFailedJsonRpcResponse,
+  parseJsonRpcResponse,
+} from "../utils/json-rpc.js";
+import { getHardhatVersion } from "../utils/package.js";
+
+import { ProviderError, ProviderErrorCode } from "./errors.js";
+
+const TOO_MANY_REQUEST_STATUS = 429;
+const MAX_RETRIES = 6;
+const MAX_RETRY_WAIT_TIME_SECONDS = 5;
+
+export class HttpProvider extends EventEmitter implements EIP1193Provider {
+  readonly #url: string;
+  readonly #networkName: string;
+  readonly #extraHeaders: Record<string, string>;
+  readonly #dispatcher: Dispatcher;
+  #nextRequestId = 1;
+
+  /**
+   * Creates a new instance of `HttpProvider`.
+   */
+  public static async create({
+    url,
+    networkName,
+    extraHeaders = {},
+    timeout,
+  }: {
+    url: string;
+    networkName: string;
+    extraHeaders?: Record<string, string>;
+    timeout: number;
+  }): Promise<HttpProvider> {
+    if (!isValidUrl(url)) {
+      throw new HardhatError(HardhatError.ERRORS.NETWORK.INVALID_URL, {
+        value: url,
+      });
+    }
+
+    const dispatcher = await getHttpDispatcher(url, timeout);
+
+    const httpProvider = new HttpProvider(
+      url,
+      networkName,
+      extraHeaders,
+      dispatcher,
+    );
+
+    return httpProvider;
+  }
+
+  /**
+   * @private
+   *
+   * This constructor is intended for internal use only.
+   * Use the static method {@link HttpProvider.create} to create an instance of
+   * `HttpProvider`.
+   */
+  constructor(
+    url: string,
+    networkName: string,
+    extraHeaders: Record<string, string>,
+    dispatcher: Dispatcher,
+  ) {
+    super();
+
+    this.#url = url;
+    this.#networkName = networkName;
+    this.#extraHeaders = extraHeaders;
+    this.#dispatcher = dispatcher;
+  }
+
+  /**
```

Comment:
I think it'd be better to place this doc in the interface, as that's public, while this is private

Reply:
Done [here](https://github.com/NomicFoundation/hardhat/pull/5580/commits/7068a5fc633606e278496d74379d8276341d11e7)